modules/microvm/virtualization/microvm/host/storage.nix

```diff
@@ -0,0 +1,115 @@
+# Copyright 2022-2024 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: Apache-2.0
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}: let
+  cfg = config.ghaf.services.storage;
+  inherit (builtins) filter map listToAttrs;
+  inherit (lib) mkIf optionals optionalAttrs optionalString forEach nameValuePair recursiveUpdate concatStringsSep unique foldl';
+
+  # Helper functions
+  filterHostShares = filter (s: s.tag != null && s.host-path != null) cfg.shares;
+  vmShares = filter (s: s.target-vm != null) cfg.shares;
+  filterSharesByName = vm: filter (s: s.target-vm == vm) vmShares;
+  sharePaths = map (s: s.host-path) filterHostShares;
+  vmList = unique (map (s: s.target-vm) vmShares);
+  mergeAttrSets = attrSets: foldl' (a: b: recursiveUpdate a b) {} attrSets;
+
+  # Share configuration
+  shareConfigs = vm: {
+    microvm.shares = forEach (filterSharesByName vm) (
+      vmShare: {
+        tag = "${vmShare.tag}";
+        source = "${vmShare.host-path}";
+        mountPoint = "${
+          if (vmShare.tmp-path != null)
+          then vmShare.tmp-path
+          else vmShare.vm-path
+        }";
+        proto = "virtiofs";
+      }
+    );
+  };
+
+  # Service configuration
+  serviceConfigs = vm:
+    forEach (filterSharesByName vm) (vmShare: {
+      environment.systemPackages = optionals (vmShare.tmp-path != null) [pkgs.umount];
+      systemd.services."storage-prep-${vmShare.tag}" = let
+        # Service to set owner, permissions, and (optionally) copy files
+        prepStorage = pkgs.writeShellScriptBin "storage_prep" ''
+          set -xeuo pipefail
+          if [ ! -d "${vmShare.vm-path}" ]; then
+            mkdir -p ${vmShare.vm-path}
+          fi
+          ${optionalString (vmShare.tmp-path != null) "cp -r ${vmShare.tmp-path}/* ${vmShare.vm-path}"}
```

Comment:
The way I understand the semantic, I would expect the previous content to be removed.
Maybe using `rsync` with `--delete` would be than a better choice?

Reply:
Rsync, also would have an option to directly apply the right owner/group without an additional chown.

Reply:
The idea with this was that the host data remains present. This 'temporary' mountpoint option is intended for the use case where one wants to copy data from the host on VM boot, and apply specific permissions without changing the host permissions or rely on some UID mapping. Deletion could be added as an option as well.

Reply:
I was also looking into using the impermanence module instead of this script. At the moment, I am waiting for the storage vm implementation to see how to finally implement this

Reply:
As far as I know, impermanence doesn't provide copy semantics.